### ISSUE TITLE
parallel MultiFormatWriter seems to "overlap"

### ISSUE THREAD

**Issue by farfromrefug** [zxing-cpp/zxing-cpp#791]
I am using `MultiFormatWriter` to generate qrcode to present them in my mobile app.
I generate them within a thread and can be generating them in parallel.
Randomly i end up with images like those below. It feels like they are using the same "buffer" as i see part of other qrcodes in some images.
I use a different `MultiFormatWriter` each time as you can see here:
```cpp
cv::Mat generateQRCode(std::string text, std::string format, int width, int height, std::string options)
{
    auto writer = ZXing::MultiFormatWriter(ZXing::BarcodeFormatFromString(format));
    if (!options.empty()) {
        jsoncons::json j = jsoncons::json::parse(options);

        if (j.contains("encoding")) {
            writer = writer.setEncoding(ZXing::CharacterSetFromString(j["encoding"].as<string>()));
        }
        if (j.contains("ecclevel")) {
            writer = writer.setEccLevel(j["ecclevel"].as<int>());
        }
        if (j.contains("margin")) {
            writer = writer.setMargin(j["margin"].as<int>());
        }
    }
    auto matrix = writer.encode(text, width, height);
    auto bitmap = ToMatrix<uint8_t>(matrix);
    const unsigned char *buffer = bitmap.data();
    cv::Mat resultMat(height,width,CV_8UC1,(unsigned char*)buffer);
    return resultMat;
}
```
Do you have any idea of what might be happening?
![Screenshot_1716989551](https://github.com/zxing-cpp/zxing-cpp/assets/655344/d04219d9-f262-47dd-9723-b64fdbc9f9da)

![Screenshot_1716989549](https://github.com/zxing-cpp/zxing-cpp/assets/655344/d7af7c6e-3060-40eb-955d-d0ce64d06179)

**Comment by axxel** [comment]
There might be some multi-threading issue with the `MultiFormatWriter` code. But since this code is about to be deprecated. Can I interest you in giving the new API and implementation a try? If you find an issue there, I'm much more inclined to spend time on it. For some quick info on how to get started, please [here](https://github.com/zxing-cpp/zxing-cpp/discussions/739#discussioncomment-8660387). You need the current HEAD revision and you need to pass `-DZXING_WRITERS=NEW -DZXING_USE_BUNDLED_ZINT -DZXING_EXPERIMENTAL_API=ON -DCMAKE_CXX_STANDARD=20` to `cmake`.

**Comment by farfromrefug** [comment]
@axxel perfect i ll test that !. On that note as we are talking about building latest. I build zxing inside an android library using this in a CMakeLists.txt
```
set(ZXING_SRC ./zxingcpp/core)
set(CMAKE_CXX_STANDARD 20)
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-deprecated-anon-enum-enum-conversion")

# ZXing
set(BUILD_READERS ON)
set(BUILD_WRITERS ON)

add_subdirectory(${ZXING_SRC} ZXing EXCLUDE_FROM_ALL)
```
(as you can see i only use core).
It worked great while using zxing master branch from february. Now i upgraded to latest and my build fails with a gradle issue. I suspected it was including/building the android wrapper.
I tried to remove the wrapper directory and then it builds. Would you have any idea on maybe new flags in the core CMakeLists.txt which could cause that? I had a quick look but for now i did not find anything

**Comment by axxel** [comment]
There is no link from the core to any of the wrapper directories. There might have been a change in the android wrapper that causes your gradle build to fail now but did not earlier. I'm not proficient enough with gradle. I would think you should be able to setup your own gradle project in a way that some other gradle project in some random subdirectory does not influence your build. Maybe gradle simply scans all subdirectories and you have to prohibit that for the wrapper directory?

**Comment by axxel** [comment]
Your `ZXing` block should now look like this
```cmake
set(ZXING_READERS ON)
set(ZXING_WRITERS NEW)
set(ZXING_USE_BUNDLED_ZINT ON)
set(ZXING_EXPERIMENTAL_API ON)
```

**Comment by farfromrefug** [comment]
@

> There is no link from the core to any of the wrapper directories. There might have been a change in the android wrapper that causes your gradle build to fail now but did not earlier. I'm not proficient enough with gradle. I would think you should be able to setup your own gradle project in a way that some other gradle project in some random subdirectory does not influence your build. Maybe gradle simply scans all subdirectories and you have to prohibit that for the wrapper directory?

Ok got it. Strange this is i dont use the android wrapper! only the cpp part of the core. Will look at it.



> Your `ZXing` block should now look like this
> 
> ```cmake
> set(ZXING_READERS ON)
> set(ZXING_WRITERS NEW)
> set(ZXING_USE_BUNDLED_ZINT ON)
> set(ZXING_EXPERIMENTAL_API ON)
> ```

Thanks i am trying with that. One issue i see with the new API: i cant set a width/height on the generated image? I only see a size hint. But i would need to enforce the image size. My use case is to generate an image which will fix screen base on device orientation. (not important with qrcode, but it is with barcodes so that it takes the whole screen in landscape).

**Comment by farfromrefug** [comment]
@axxel so i tested `ZXING_EXPERIMENTAL_API` but it wont build on android :s seems like it uses std features not available on android(specifically std::ranges). And i also get the error `initialization of incomplete type 'const ZXing::WriterOptions'`.
Not sure if i can fix that. I have been testing quite a lot of things without success

**Comment by axxel** [comment]
> One issue i see with the new API: i cant set a width/height on the generated image? I only see a size hint.

Yes. And that is deliberate. If you need a specific size, you can always add whitepace yourself. The precise reasoning can be found somewhere in the lengthy discussion of this new API.

Your mentioning of the SVG API suggests that went that route, so you don't need pixels anymore? That should generally be considered the best option, anyway.

> EDIT: it builds with NDK 26.1.10909125!

So just upgrading your NDK fixed all build issues?

**Comment by farfromrefug** [comment]
> > One issue i see with the new API: i cant set a width/height on the generated image? I only see a size hint.
> 
> Yes. And that is deliberate. If you need a specific size, you can always add whitepace yourself. The precise reasoning can be found somewhere in the lengthy discussion of this new API.
> 
> Your mentioning of the SVG API suggests that went that route, so you don't need pixels anymore? That should generally be considered the best option, anyway.
> 
> > EDIT: it builds with NDK 26.1.10909125!
> 
> So just upgrading your NDK fixed all build issues?

Yes it seems you need NDK 26+ even if you already use c++20. I think it is because it comes with a newer version of clang which support a bigger range of c++20 features

**Comment by axxel** [comment]
Does the switch to the new API / zint-backend fix your problem?

**Comment by farfromrefug** [comment]
@axxel sorry for not replying sooner yes never had that issue anymore with the new AP!
